ecl/hql/hqlutil.cpp

```diff
@@ -8832,47 +8853,122 @@ StringBuffer & appendLocation(StringBuffer & s, IHqlExpression * location, const
 
 //---------------------------------------------------------------------------------------------------------------------
 
-static void createMappingAssigns(HqlExprArray & assigns, IHqlExpression * selfSelector, IHqlExpression * oldSelector, IHqlSimpleScope * oldScope, IHqlExpression * newRecord)
+static bool doReportDroppedFields(IHqlSimpleScope * newScope, IHqlExpression * oldRecord, IErrorReceiver &err, ECLlocation &location)
+{
+    bool allDropped = true; // until we find one that isn't
+    ForEachChild(i, oldRecord)
+    {
+        IHqlExpression * cur = oldRecord->queryChild(i);
+        switch (cur->getOperator())
+        {
+        case no_record:
+            allDropped = doReportDroppedFields(newScope, cur, err, location) && allDropped;
+            break;
+        case no_ifblock:
+            allDropped = doReportDroppedFields(newScope, cur->queryChild(1), err, location) && allDropped;
+            break;
+        case no_field:
+            {
+                OwnedHqlExpr newField = newScope->lookupSymbol(cur->queryId());
+                if (!newField)
+                {
+                    VStringBuffer msg("Field %s is present in DFS file but not in ECL definition", str(cur->queryId()));
+                    err.reportWarning(CategoryInformation, HQLINFO_FieldNotPresentInECL, msg.str(), str(location.sourcePath), location.lineno, location.column, location.position);
+                }
+                else
+                {
+                    allDropped = false;
+                    if (newField->queryType() != cur->queryType())
+                    {
+                        VStringBuffer msg("Field %s type mismatch: DFS reports ", str(cur->queryId()));
+                        cur->queryType()->getECLType(msg).append(" but ECL declared ");
+                        newField->queryType()->getECLType(msg);
+                        err.reportWarning(CategoryDFS, HQLWRN_DFSlookupTypeMismatch, msg.str(), str(location.sourcePath), location.lineno, location.column, location.position);
+                    }
+                }
+            }
+        }
+    }
+    return allDropped;
+}
+
+void reportDroppedFields(IHqlExpression * newRecord, IHqlExpression * oldRecord, IErrorReceiver &err, ECLlocation &location)
+{
+    if (doReportDroppedFields(newRecord->querySimpleScope(), oldRecord, err, location))
+    {
+        err.reportWarning(CategoryDFS, HQLWRN_NoFieldsMatch, "No matching fields found in ECL definition", str(location.sourcePath), location.lineno, location.column, location.position);
+    }
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+static void createMappingAssigns(HqlExprArray & assigns, IHqlExpression * selfSelector, IHqlExpression * oldSelector, IHqlSimpleScope * oldScope, IHqlExpression * newRecord, bool replaceMissingWithDefault, IErrorReceiver &err, ECLlocation &location)
```

Comment:
efficiency/clarity.  With this many parameters it might be worth converting this in the future into a member function of a class.